torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3582,6 +3671,94 @@ def guard_from(expr, concrete_val):
 
         return concrete_val
 
+    def refine_ranges(self, guard: ShapeGuard) -> None:
+        def simplify(expr: sympy.Expr) -> sympy.Expr:
+            """
+            Simplification specialized for range refinement.
+            """
+            return self.try_isolate_symbol_lhs(self.simplify(expr))
+
+        def simplify_until(expr: sympy.Expr, max_iterations: int = 10) -> sympy.Expr:
+            """
+            Calls 'simplify' either until it does not change or until it reaches the
+            maximum number of iterations.
+            """
+            for _ in range(max_iterations):
+                previous, expr = expr, simplify(expr)
+                if expr == previous:
+                    break
+            return expr
+
+        RELOP_MIRROR = {
+            sympy.Ge: sympy.Le,
+            sympy.Gt: sympy.Lt,
+            sympy.Le: sympy.Ge,
+            sympy.Lt: sympy.Gt,
+        }
+
+        # List of expressions to be processed.
+        # Here, we try considering both LHS and RHS by mirroring the
+        # original expression: a < b ==> b > a
+        exprs = [guard.expr]
+
+        if type(guard.expr) in RELOP_MIRROR:
+            exprs.append(RELOP_MIRROR[type(guard.expr)](guard.expr.rhs, guard.expr.lhs))  # type: ignore[arg-type]
+
+        for expr in exprs:
+            # First, try to simplify the left-hand side.
+            expr = simplify_until(expr)
+
+            # Filter the guards that:
+            #   1. are relational operations
+            #   2. have a symbol as the left-hand side
+            #   3. already have a range
+            if (
+                not isinstance(expr, sympy.Rel)
+                or not isinstance(expr.lhs, sympy.Symbol)
+                or expr.lhs not in self.var_to_range
+            ):
```

Comment:
nit, we're essentially doing a pattern match here, so to me it would be more clear to demorganize into not (X and Y) to show that afterwards X and Y are true.